src/wrappers/themis/wasm/src/secure_cell_context_imprint.js

```diff
@@ -0,0 +1,190 @@
+        message = utils.coerceToBytes(message)
+        if (message.length == 0) {
+            throw new ThemisError(subsystem, ThemisErrorCode.INVALID_PARAMETER,
+                'message must be not empty')
+        }
+
+        // Other Secure Cell kinds have optional context, so the users are likely to omit
+        // the context here as well. Let's produce a more helpful error message instead
+        // of an error about 'undefined' being invalid byte buffer type.
+        if (context === undefined) {
+            throw new ThemisError(subsystem, ThemisErrorCode.INVALID_PARAMETER,
+                'SecureCellContextImprint requires context')
+        }
+
+        context = utils.coerceToBytes(context)
+        if (context.length == 0) {
+            throw new ThemisError(subsystem, ThemisErrorCode.INVALID_PARAMETER,
+                'context must be not empty')
+        }
+
+        let status
+        /// C API uses "size_t" for lengths, it's defined as "i32" in Emscripten
+        let result_length_ptr = libthemis.allocate(4, 'i32', libthemis.ALLOC_STACK)
+        let master_key_ptr, message_ptr, context_ptr, result_ptr, result_length
+        try {
+            master_key_ptr = libthemis._malloc(this.masterKey.length)
+            message_ptr = libthemis._malloc(message.length)
+            context_ptr = libthemis._malloc(context.length)
+            if (!master_key_ptr || !message_ptr || !context_ptr) {
+                throw new ThemisError(subsystem, ThemisErrorCode.NO_MEMORY)
+            }
+
+            libthemis.writeArrayToMemory(this.masterKey, master_key_ptr)
+            libthemis.writeArrayToMemory(message, message_ptr)
+            libthemis.writeArrayToMemory(context, context_ptr)
+
+            status = libthemis._themis_secure_cell_encrypt_context_imprint(
+                master_key_ptr, this.masterKey.length,
+                message_ptr, message.length,
+                context_ptr, context.length,
+                null, result_length_ptr
+            )
+            if (status != ThemisErrorCode.BUFFER_TOO_SMALL) {
+                throw new ThemisError(subsystem, status)
+            }
+
+            result_length = libthemis.getValue(result_length_ptr, 'i32')
+            result_ptr = libthemis._malloc(result_length)
+            if (!result_ptr) {
+                throw new ThemisError(subsystem, ThemisErrorCode.NO_MEMORY)
+            }
+
+            status = libthemis._themis_secure_cell_encrypt_context_imprint(
+                master_key_ptr, this.masterKey.length,
+                message_ptr, message.length,
+                context_ptr, context.length,
+                result_ptr, result_length_ptr
+            )
+            if (status != ThemisErrorCode.SUCCESS) {
+                throw new ThemisError(subsystem, status)
+            }
+
+            result_length = libthemis.getValue(result_length_ptr, 'i32')
+
+            return libthemis.HEAPU8.slice(result_ptr, result_ptr + result_length)
```

Comment:
voodoo magic 🙈

Reply:
Not so much. `libthemis.HEAPU8` is an unsigned byte array representing all heap memory of Emscripten. Indexing this array is the same as dereferencing a pointer. `slice` is an array method which returns an array with copy of elements between provided indices. Effectively, this is a counterpart of `writeArrayToMemory` function that copies an array of data back from Emscripten memory to JavaScript memory.